trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java

```diff
@@ -2844,7 +2844,11 @@ public ExpressionResult makeAssignment(final ILocation loc, final LRValue leftHa
 		final AssignmentStatement assignStmt = StatementFactory.constructAssignmentStatement(loc,
 				new LeftHandSide[] { lValue.getLhs() }, new Expression[] { rhsWithBitfieldTreatment });
 
-		builder.addStatement(assignStmt);
+		if (leftHandSide.getCType().isAtomic()) {
+			builder.addStatement(new AtomicStatement(loc, new Statement[] { assignStmt }));
+		} else {
+			builder.addStatement(assignStmt);
+		}

```

Comment:
Note that compound assignment expressions such as `x += 3` also have to be treated atomically, as per paragraph 3 in Section 6.5.12.2 of the C standard (N1570):

> ... If E1 has an atomic type, compound assignment is a read-modify-write operation with memory_order_seq_cst memory order semantics.

Reply:
Thanks for the hint. For non-heap variables this actually works as it is (as the read does not lead to additional statements), but it is currently handled unsound for heap variables (as the translation leads to two separate atomic blocks).

Reply:
Does this (for off-heap variables) also work correctly (1) with the "no-data-race LBE" setting and (2) when data race checks are added (which would be additional statements for the read) ?

Reply:
Should be fixed in [a2a635d](https://github.com/ultimate-pa/ultimate/pull/688/commits/a2a635d79fe1f17aaf6bb23bb30828a10d4b7d5a)

Reply:
Do we still need the handling in this method?

Reply:
Yes, this method is to handle simple assignment. For clarification I will rename the method `makeAtomicAssignment` to `handleAtomicReadWrite`, since it is used for this purpose.